### ISSUE TITLE
overlap fails when other pyrange is empty

### ISSUE THREAD

**Issue by tzuni** [biocore-ntnu/pyranges#4]
Hi, 

I ran into this problem when finding what ranges from one set overlap ranges in several other sets. Here is a trivial test case:

```python
pr1 = pr.PyRanges(pd.DataFrame({
    'Chromosome': ['1', '1'],
    'Start': [243, 645],
    'End': [267, 689],
}))
pr2 = pr.PyRanges(pd.DataFrame({
    'Chromosome': ['2', '2'],
    'Start': [243, 645],
    'End': [267, 689],
}))
pr3 = pr.PyRanges(pd.DataFrame({
    'Chromosome': ['3', '3'],
    'Start': [243, 645],
    'End': [267, 689],
}))

# overlap of pr1 and pr2; should be empty
s1 = pr1
s2 = pr2
res1 = s1.overlap(s2, strandedness=None, how='first')

# overlap of first comparison with pr3; should also be empty
s1 = res1
s2 = pr2
res2 = s1.overlap(s2, strandedness=None, how='first')
```
Which generates this error:

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-158-047be8f5ac7d> in <module>
     17 s1 = res1
     18 s2 = pr2
---> 19 res2 = s1.overlap(s2, strandedness=None, how='first')
     20 len(res2.df)

~/jupyter-venv/lib/python3.7/site-packages/pyranges/pyranges.py in overlap(self, other, **kwargs)
    423         kwargs = fill_kwargs(kwargs)
    424 
--> 425         dfs = pyrange_apply(_overlap, self, other, **kwargs)
    426 
    427         return PyRanges(dfs)

~/jupyter-venv/lib/python3.7/site-packages/pyranges/multithreaded.py in pyrange_apply(function, self, other, **kwargs)
    215     else:
    216 
--> 217         if self.stranded and not other.stranded:
    218 
    219             for (c, s), df in items:

~/jupyter-venv/lib/python3.7/site-packages/pyranges/pyranges.py in stranded(self)
    686     def stranded(self):
    687         # print(self.keys())
--> 688         return len(list(self.keys())[0]) == 2
    689 
    690     @property

IndexError: list index out of range
```

The immediate problem is since res1 is empty it doesn't have anything with which to populate the dataframe dictionary `dfs`. So when checking if either set is stranded it fails because `self.keys()` is empty

```
    @property
    def stranded(self):
        return len(list(self.keys())[0]) == 2
```

Not sure if this is the best fix, but `stranded(self)` should probably first check if `len(self.keys()) > 0` or something. Alternately, maybe just save a lot of computational work and check in overlap if either `PyRanges` is empty and if so return an empty `PyRanges` object.

**Comment by endrebak** [comment]
Wow, thanks for the detective work. I should think up some hypothesis tests that do three or more operations in a row. I wouldn’t need to check the result really, just that no exception is thrown :)

**Comment by endrebak** [comment]
Dunno how much you care about internals, but: Pyranges has a function dispatcher which is the same for all binary functions. It takes care of the strand logic, multithreading and so on. That is actually the place where this is happening, before the overlap function is dispatched. Pyranges is extremely thoroughly tested with automatically generated tests, but it is always tested with pyranges of at least one row. I’ll create some tests where zero row pyranges are allowed also :)

**Comment by endrebak** [comment]
Btw, this is syntax you can use to create an empty gr:

`gr_empty = pr.PyRanges(pd.DataFrame(columns="Chromosome Start End".split()))`

Will try to fix this tonight after sending in the paper. This raises the question whether empty pyranges should be considered stranded or not 🤔

**Comment by tzuni** [comment]
Awesome! Glad to help! I'm very excited about this project. It's always the case that users almost immediately find cases that break your painstakingly polished code.

That is shorter than what I'd been using:
```
pr.PyRanges(pd.DataFrame({
            'Chromosome': [],
            'Start': [],
            'End': [],
        }))
```
It seems like the problem with mixing stranded and non-stranded features together stems from the difficulty in combining DataFrames with optional 'Strand' columns. Maybe the 'Strand' columns should be required and a third symbol (like `*` or `.`) used to indicate it's lack of application to that feature. They could be automatically added and set to the no strand symbol if missing. What do you think? I'm not sure of all the implications here.

Also, as part what I was working on I created a fairly clean solution to check for ranges from one set that overlap ranges in all members of a list of other ranges. Would you be open to accepting a pull-request for code that adds this and similar functionality?

**Comment by endrebak** [comment]
I'm hesitant to add code and want to keep the core library fairly slim. It is better to create additional libraries I think. I have plenty of useful functionality that I have kept out of pyranges, but which I will release as other packages eventually. Tidy-functions, GenomicFeatures-functions and so on :)

Does your function do something like the following?

```
pr.concat([query_range.overlap(subject_range) for subject_range in subject_ranges])
```

**Comment by endrebak** [comment]
If a PyRange is empty, it is not technically correct to call it stranded or unstranded. The main problem is that most functionality should work with empty PyRanges, i.e. `empty_pyrange.intersect(other_pyrange, strandedness="same")` should return a valid result, not err with a `cannot perform stranded operations on unstranded PyRanges`-error :)

Will think more about strandedness, I guess that is a discussion for a different thread.

**Comment by tzuni** [comment]
Sounds good.

So if I have a list of PyRanges: `[A, B, C, D]` I needed a list of all the ranges in A that overlapped a range in ALL of `B, C, D`. This is the code I have:

```python
from functools import reduce

def multioverlap(iv1, iv2):
    return iv1.overlap(iv2, strandedness=None, how='first')

# assume A, B, C, D are existing PyRanges objects
interval_list = [A, B, C, D]
ovlp = reduce(multioverlap, interval_list)
# ovlp now has ranges from A which overlap ranges in all of B, C, D
```

**Comment by endrebak** [comment]
Then this is what you want:

```
pr.concat([A.overlap(subject_range, how="first") for gr in [B, C, D]])
```

The `pr.concat` reduces all the pyranges to one :) Thanks for explaining.

**Comment by tzuni** [comment]
Wouldn't that give all the ranges in A that overlap ANY range from B, C, D ?

**Comment by endrebak** [comment]
Ah, sorry, yes ignore. I am too tired. I love that you can do that so tersely in Python :) But because of that, I feel that it does not need to be a part of the library. But it would be good to include in a tutorial/book on PyRanges.

One-liner:

```
reduce(lambda iv1, iv2: iv1.overlap(iv2, strandedness=None, how='first'), interval_list)
```

**Comment by tzuni** [comment]
No worries, I know the feeling! Thanks for your work on this, I look forward to using it extensively!

**Comment by endrebak** [comment]
This is fixed in `0.0.19`. Thanks for reporting. There are other bugs that happen when you chain a lot of data, working on fixing/testing this with https://github.com/biocore-ntnu/pyranges/blob/master/tests/test_do_not_error.py

Will continue after the weekend.
